### PR TITLE
Robustify HIL Tests

### PR DESCRIPTION
**Pull request by matt001k**
## What changed?
Adds retry logic to HIL tests
Adds topology helper script CLI option if script invoked directly
Fixes neighbors test (previously would check each port and verify that a neighbor is associated with each port even if there were not two neighbors or a single neighbor was on the opposite port)

## How does it make Bristlemouth better?
Adds the ability to retry tests just in case a serial message was missed,
can invoke topology script directly to obtain a systems topology (useful for bm_protocol's HIL implementation run on a local runner)

## Where should reviewers focus?
Mainly the retry logic, ensure it looks adequate for the tests.

## Checklist

- [ ] Add or update unit tests for changed code
- [ ] Ensure all submodules up to date. If this PR relies on changes in submodules, merge those PRs first, then point this PR at/after the merge commit
- [ ] Ensure code is formatted correctly with clang-format. If there are large formatting changes, they should happen in a separate whitespace-only commit on this PR after all approvals.

### CODE REVIEW

File: test/scripts/test_ping.py [hunk 1]
```diff
@@ -1,13 +1,24 @@
...
-from util import format_node_id_to_hex_str
+from util import format_node_id_to_hex_str, retry_test
 
 
 class TestPing:
     """Ping test class"""
 
     TEST_STRING = "test_string"
 
+    def __compare_data(self, ser: SerialHelper):
```
Comment (towynlin) [inline test/scripts/test_ping.py:11]: I'm curious why this function has a leading double underscore. Dunderscore causes some specific class-private (not in subclasses) name-mangling behavior.

If there's not some more specific reason than signaling that this is private, use a single underscore, or even none, since there wouldn't inherently be a problem with calling this method externally.

File: test/scripts/topology_helper.py [hunk 2]
```diff
@@ -61,3 +62,58 @@ def root(self) -> int:
...
+    baud = None
+    baud_arg = False
+
+    for arg in sys.argv:
+        if arg == "--help" or arg == "-h":
+            help = True
+            break
+        elif arg == "--root" or arg == "-r":
+            root = True
+        elif arg == "--port" or arg == "-p":
```
Comment (towynlin) [inline test/scripts/topology_helper.py:86]: Let's replace this with argparse: https://docs.python.org/3/library/argparse.html It generates the help text in a standard way automatically and handles more edge cases than we'll think of, plus it'll reduce this code a lot.

File: test/scripts/util.py [hunk 3]
```diff
@@ -38,3 +41,56 @@ def print_progress_bar(s: str, current: Union[int, float], total: Union[int, flo
...
+                except serial.SerialException as serial_error:
+                    serial_message = serial_error.__str__().split("\n")[0]
+                    print(
+                        f'Retry error: "{test_fun.__name__}" '
+                        f"--> {serial_message}."
+                        f"[{retry_count}/{max_attempts - 1}]"
+                        f" Retrying new execition in {wait_s} second(s)"
+                    )
+                    time.sleep(wait_s)
+                    retry_count += 1
```
Comment (towynlin) [inline test/scripts/util.py:89]: Logic looks good. Cool use of a decorator! Since the error handling is identical between the 2 cases can we do `except (AssertionError, serial.SerialException) as err` and avoid duplicating the 9 lines of print-wait-increment code?

**Review by towynlin — CHANGES_REQUESTED** [review]

File: test/scripts/test_ping.py (hunk 1 above)
Comment (matt001k) [reply test/scripts/test_ping.py:11]: This was just to indicate that this is private. I will update.

**Review by towynlin — APPROVED** [review]

**Review by victorsowa12 — APPROVED** [review]
